[docdb] Tool to trigger a flush or compaction on the sys_catalog tablet

Tigojiang opened the pull request:
pull request for issue 8730
https://github.com/yugabyte/yugabyte-db/issues/8730

bmatican commented:
@Tigojiang thanks for the PR! I'll let @robertsami take a pass

meantime, 2 high level asks
- any chance you can get this to also do compactions? a quick cursory look shows you only glued in the flush capabilities
- can you also add a note in the PR about how you tested this? ok to edit the description to capture the test plan

Tigojiang commented:
> @Tigojiang thanks for the PR! I'll let @robertsami take a pass
> 
> meantime, 2 high level asks
> 
> * any chance you can get this to also do compactions? a quick cursory look shows you only glued in the flush capabilities
> * can you also add a note in the PR about how you tested this? ok to edit the description to capture the test plan
Hi, I have added the test plan. Oh is compaction another command to add? What is the exact command name so I can add on it? Currently, I know the one that I have added is flush_sys_catalog. Thank you!

robertsami commented:
> Hi @bmatican, I have added the test plan. Oh is compaction another command to add? What is the exact command name so I can add on it? Currently, I know the one that I have added is flush_sys_catalog. Thank you!

lets add a compact_sys_catalog command as well!

robertsami commented on src/yb/master/catalog_manager.h:
```diff
@@ -261,6 +261,11 @@ class CatalogManager :
                                      IsTruncateTableDoneResponsePB* resp);
 
+  // Flush the system catalog
+  CHECKED_STATUS FlushSysCatalog(const FlushSysCatalogRequestPB* req, 
+                                 FlushSysCatalogResponsePB* resp, 
+                                 rpc::RpcContext* rpx);
```
```suggestion
  CHECKED_STATUS FlushSysCatalog(const FlushSysCatalogRequestPB* req,
                                 FlushSysCatalogResponsePB* resp,
                                 rpc::RpcContext* rpx);
```

robertsami commented on src/yb/master/master_service.h:
```diff
@@ -274,6 +274,10 @@ class MasterServiceImpl : public MasterServiceIf,
                    FlushTablesResponsePB* resp,
                    rpc::RpcContext rpc) override;
 
+  void FlushSysCatalog(const FlushSysCatalogRequestPB* req, 
+                       FlushSysCatalogResponsePB* resp,
+                       rpc::RpcContext rpx) override;
```
```suggestion
  void FlushSysCatalog(const FlushSysCatalogRequestPB* req,
                       FlushSysCatalogResponsePB* resp,
                       rpc::RpcContext rpx) override;
```

robertsami commented on src/yb/tools/yb-admin-test.cc:
```diff
@@ -714,6 +714,11 @@ TEST_F(AdminCliTest, DdlLog) {
   ASSERT_EQ(actions[2], "Add column int_column[int32 NULLABLE NOT A PARTITION KEY]");
 }
-
+TEST_F(AdminCliTest, FlushSysCatalog) {
+  BuildAndStart();
+  string master_address = ToString(cluster_->master()->bound_rpc_addr());
+  auto client = ASSERT_RESULT(YBClientBuilder().add_master_server_addr(master_address).Build());
+  ASSERT_OK(CallAdmin("flush_sys_catalog", 20));
+}
```
```suggestion

TEST_F(AdminCliTest, FlushSysCatalog) {
  BuildAndStart();
  string master_address = ToString(cluster_->master()->bound_rpc_addr());
  auto client = ASSERT_RESULT(YBClientBuilder().add_master_server_addr(master_address).Build());
  ASSERT_OK(CallAdmin("flush_sys_catalog", 20));
}

```

robertsami commented on src/yb/tools/yb-admin_cli.cc:
```diff
@@ -526,6 +526,19 @@ void ClusterAdminCli::RegisterCommandHandlers(ClusterAdminClientClass* client) {
+  Register(
+      "flush_sys_catalog", 
+      "[timeout_in_seconds] (default 20)",
+      // " <table> [timeout_in_seconds] (default 20)"
```
```suggestion
```

robertsami commented on src/yb/tools/yb-admin_cli.cc:
```diff
@@ -526,6 +526,19 @@ void ClusterAdminCli::RegisterCommandHandlers(ClusterAdminClientClass* client) {
+        int timeout_secs = 20;
+        // We use the FlushSysCatalog RPC to trigger compaction.
+        RETURN_NOT_OK_PREPEND(
+            client->FlushSysCatalog(timeout_secs /* is_compaction */),
```
```suggestion
            client->FlushSysCatalog(timeout_secs),
```

robertsami commented on src/yb/tools/yb-admin_cli.cc:
```diff
@@ -526,6 +526,19 @@ void ClusterAdminCli::RegisterCommandHandlers(ClusterAdminClientClass* client) {
+      // " <table> [timeout_in_seconds] (default 20)"
+      [client](const CLIArguments& args) -> Status {
+        int timeout_secs = 20;
+        // We use the FlushSysCatalog RPC to trigger compaction.
```
```suggestion
```

robertsami commented on src/yb/tools/yb-admin_cli.cc:
```diff
@@ -526,6 +526,19 @@ void ClusterAdminCli::RegisterCommandHandlers(ClusterAdminClientClass* client) {
+      "[timeout_in_seconds] (default 20)",
+      // " <table> [timeout_in_seconds] (default 20)"
+      [client](const CLIArguments& args) -> Status {
+        int timeout_secs = 20;
```
how do we get the user-provided timeout?

Tigojiang replied:
@robertsami  Hi, for the flush_table, it uses VERIFY_RESULT(GetTimeoutAndAddIndexesFlag(
          args.begin() + 1, args.end())); to bind timeout_secs and add_index. I can use the same one for binding. Then do I need to add add_indexes for flush_sys_catalog? What is add_indexes used for?

robertsami commented on src/yb/master/master.proto:
```diff
@@ -1996,6 +1996,14 @@ message FlushTablesResponsePB {
+message FlushSysCatalogResponsePB{
+  optional MasterErrorPB error = 1;
+
+  optional bytes flush_request_id = 2;
```
do we use this field anywhere?

Tigojiang replied:
Actually not yet. I look at the codes for FlushTablesResponsePB, which has these two fields and at first write out the same.

Tigojiang replied:
After fix the code in catalog_manager.cc to return the status, these fields might be used

robertsami commented on src/yb/master/catalog_manager.cc:
```diff
@@ -4046,6 +4046,14 @@ Status CatalogManager::GetBackfillJobs(
   return Status::OK();
 }
 
+Status CatalogManager::FlushSysCatalog(
+    const FlushSysCatalogRequestPB* req,
+    FlushSysCatalogResponsePB* resp,
+    rpc::RpcContext* rpc) {
+  RETURN_NOT_OK(sys_catalog_->tablet_peer()->shared_tablet()->Flush(tablet::FlushMode::kSync));
+  return Status::OK();
+}
```
we need to return/populate an error in the response in case this flush operation fails. see GetBackfillJobs just above this for an example:
```
if (indexed_table == nullptr) {
    Status s = STATUS(NotFound, "Requested table $0 does not exist", table_id.ShortDebugString());
    return SetupError(resp->mutable_error(), MasterErrorPB::OBJECT_NOT_FOUND, s);
  }
```

Tigojiang replied:
will return_not_ok populate the error from Flush(tablet::FlushMode::kSync) back as a bad status?

robertsami replied:
@Tigojiang yes!

robertsami reviewed (changes requested):
nice work @Tigojiang! i suggested some cosmetic changes based on our coding style and added a couple of minor points of feedback as well. a few high level points otherwise:
1. can you make this PR from a clean branch in your forked repo without merge commits? it will make it easier for me to pull your changes locally and run them through our CI confidently (see https://gist.github.com/Chaser324/ce0505fbed06b947d962 for reference)
2. for your test plan, can you include some steps where you more explicitly check that a flush actually happened? i can see two ways to do this:
    a. check the sizes of memtables before and after you flush to confirm the flush happened correctly. i think you can look at http://localhost:7000/mem-trackers and "Current Consumption" of "RegularDB-00000000000000000000000000000000" drop before vs. after your flush. ping me if this doesn't work out easily
    b. maybe _easier_ you may instead want to look at the sys catalog directory on your local machine and just ensure new SSTs are created when you flush (my directory is at `/Users/rob/var/data/yb-data/master/data/rocksdb/table-sys.catalog.uuid/`, yours should be located similarly if you run yugabyte with `bin/yugabyted start`)
3. just reiterating, lets add a compact_sys_catalog command as well, and please add this to your test plan. you can test this by flushing the sys_catalog multiple times, noting that there are multiple SSTs on disk, and then compacting them into one and verifying only one is present after

Tigojiang commented:
@robertsami Hi, after running the command, table-sys.catalog.uuid does get updated with new sst file. Is it enough for the test or do I need to check if the file exist in the unit test?
